src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -414,5 +432,95 @@ private static void CheckForCompletedTransfers(SafeCurlMultiHandle multiHandle,
                 }
             }
         }
+
+        private async static void PollFunction(SafeCurlMultiHandle multiHandle)
+        {
+            CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+            while (true)
+            {
+                if (cancellationTokenSource.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                int count = 0;
+                lock (multiHandle)
+                {
+                    multiHandle.CancellationTokenSource = cancellationTokenSource;
+                    if (null != multiHandle.FdSet)
+                    {
+                        count = multiHandle.FdSet.Count;
+                    }
+                }
+                if (count == 0)
+                {
+                    // This can happen rarely when the socket callback is not
+                    // yet invoked for the first request
+                    await Task.Delay(100).ConfigureAwait(false);
+                    continue;
+                }
+
+                int result = -1;
+                unsafe
+                {
+                    Interop.libc.pollfd* fds = stackalloc Interop.libc.pollfd[count];
+                    lock (multiHandle)
+                    {
+                        if (count != multiHandle.FdSet.Count)
+                        {
+                            continue;
+                        }
+                        int i = 0;
+                        foreach (int fd in multiHandle.FdSet)
+                        {
+                            fds[i].fd = fd;
+                            fds[i].events = PollFlags.POLLIN | PollFlags.POLLOUT;
+                            i++;
+                        }
+                    }
+
+                    var retVal = Interop.libc.poll(fds, (uint)count, -1);
```

Comment:
If we assume for now that we are going to dedicate a thread to waiting (which isn't terrible as long as the same HttpClient instance is being used for lots of concurrent operations rather than having lots of HttpClient instances each with one operation)...

How about creating your own file descriptor that you can use for artificially waking up a poll.  Then you can have a loop like (pseudo-code):
```C#
while (true)
{
    pollfd[] fds = new pollfd[Count + 1];
    for (int i = 0; i < Count; i++)
        fds[i] = ...;
    fds[fds.Length - 1] = specialWakeUpDescriptor;
    int retval = poll(pollfds, fds.Length, -1);
    if (retval == fds.Length) // special wake-up fd
        continue; // we got signaled to start over, so loop around
    ... // process the real signaled descriptor
}
```